### ISSUE TITLE
Animations don't work on scatter3d and jupyter notebook without `validate=False`

### ISSUE THREAD

**Issue by master-q** [plotly/plotly.py#948]
I tried to draw 3D animation on jupyter notebook after reading #617. But the code needs `validate=False`.

```python
# https://github.com/plotly/plotly.py/issues/617

from plotly.offline import init_notebook_mode, iplot

init_notebook_mode(connected=False)

figure_or_data = {'data': [{'x': [1, 2], 'y': [0, 1], 'z': [2, 2], 'type': 'scatter3d'}],
                  'layout': {'xaxis': {'range': [0, 3], 'autorange': False},
                             'yaxis': {'range': [0, 20], 'autorange': False},
                             'zaxis': {'range': [0, 20], 'autorange': False},
                             'title': 'First Title'},
                  'frames': [{'data': [{'x': [1, 2], 'y': [5, 7], 'z': [1, 2]}]},
                             {'data': [{'x': [-1, 3], 'y': [3, 9], 'z': [2, 2]}]},
                             {'data': [{'x': [2, 2.6], 'y': [7, 5], 'z': [3, 7]}]},
                             {'data': [{'x': [1.5, 3], 'y': [7.5, 4], 'z': [4, 7]}]},
                             {'data': [{'x': [1, 2], 'y': [0, 1], 'z': [5, 1]}],
                             'layout': {'title': 'End Title'}}]}

# Why need "validate=False"?
iplot(figure_or_data, validate=False)
```

If you run above code without `validate=False`, you will see following error messages:

```
---------------------------------------------------------------------------
PlotlyError                               Traceback (most recent call last)
<ipython-input-16-049ee65555d8> in <module>()
     18 
     19 # Why need "validate=False"?
---> 20 iplot(figure_or_data)

/home/km-okabe/.local/lib/python2.7/site-packages/plotly/offline/offline.pyc in iplot(figure_or_data, show_link, link_text, validate, image, filename, image_width, image_height, config)
    330     config.setdefault('linkText', link_text)
    331 
--> 332     figure = tools.return_figure_from_figure_or_data(figure_or_data, validate)
    333 
    334     # Though it can add quite a bit to the display-bundle size, we include

/home/km-okabe/.local/lib/python2.7/site-packages/plotly/tools.pyc in return_figure_from_figure_or_data(figure_or_data, validate_figure)
   1462                                          "plot option.\nHere's why you're "
   1463                                          "seeing this error:\n\n{0}"
-> 1464                                          "".format(err))
   1465         if not figure['data']:
   1466             raise exceptions.PlotlyEmptyDataError(

PlotlyError: Invalid 'figure_or_data' argument. Plotly will not be able to properly parse the resulting JSON. If you want to send this 'figure_or_data' to Plotly anyway (not recommended), you can set 'validate=False' as a plot option.
Here's why you're seeing this error:

'z' is not allowed in 'scatter'

Path To Error: ['frames'][0]['data'][0]['z']

Valid attributes for 'scatter' at path ['frames'][0]['data'][0] under parents ['figure', 'frames', 'frames_entry', 'data']:

    ['cliponaxis', 'connectgaps', 'customdata', 'customdatasrc', 'dx',
    'dy', 'error_x', 'error_y', 'fill', 'fillcolor', 'hoverinfo',
    'hoverinfosrc', 'hoverlabel', 'hoveron', 'hovertext', 'hovertextsrc',
    'ids', 'idssrc', 'legendgroup', 'line', 'marker', 'mode', 'name',
    'opacity', 'r', 'rsrc', 'selected', 'selectedpoints', 'showlegend',
    'stream', 't', 'text', 'textfont', 'textposition', 'textpositionsrc',
    'textsrc', 'tsrc', 'type', 'uid', 'unselected', 'visible', 'x', 'x0',
    'xaxis', 'xcalendar', 'xsrc', 'y', 'y0', 'yaxis', 'ycalendar', 'ysrc']

Run `<scatter-object>.help('attribute')` on any of the above.
'<scatter-object>' is the object at ['frames'][0]['data'][0]
```

The message said "z is not allowed", but I choose `scatter3d`...

How to fix it with good breeding?

**Comment by master-q** [comment]
I'm sorry, following works:

```python
import plotly.graph_objs as go
from plotly.offline import init_notebook_mode, iplot

init_notebook_mode(connected=False)

figure = {'data': [{'x': [0, 1], 'y': [0, 1], 'z': [0, 1], 'type': 'scatter3d'}],
          'layout': {'scene': {'xaxis': {'range': [0, 5], 'autorange': False},
                               'yaxis': {'range': [0, 5], 'autorange': False},
                               'zaxis': {'range': [0, 5], 'autorange': False}},
                     'updatemenus': [{'type': 'buttons',
                                      'buttons': [{'label': 'Play',
                                                   'method': 'animate',
                                                   'args': [None]}]}]
                    },
          'frames': [{'data': [{'x': [1, 2], 'y': [1, 2], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [1, 4], 'y': [1, 4], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [3, 4], 'y': [3, 4], 'z': [0, 3], 'type': 'scatter3d'}]}]
         }

iplot(figure)
```

**Comment by master-q** [comment]
BTW, the `autorange` doesn't work. Where should I put the parameter?

**Comment by fperez** [comment]
Can confirm... 

cc @jasongrout @gnestor: any ideas guys? Is this limitation related to the (blocked) display of JS in Lab?  In Lab, the above example produces this:

![image](https://user-images.githubusercontent.com/57394/37445031-7cdfa948-27d2-11e8-9f94-e10a72b2513f.png)

but the 'Play' button is dead. In Classic, the button animates through the provided frames.  Any thoughts?

**Comment by jasongrout** [comment]
@fperez - it seems there are several conflated issues here? 1. The original issue seemed to be an error in the original code. That seems to be corrected? 2. There was another question about autorange, but that also seems to be a question about how to write a plotly plot. And both reference the notebook - @master-q: do you mean it doesn't work in the classic notebook, or in JupyterLab?

Then @fperez, in your comment you're asking about things working in jlab. Did you see an error message printed to the console about not executing js? (https://github.com/jupyterlab/jupyterlab/blob/8de8cf6d361329fe975627d09a0cc12a517eec4a/packages/rendermime/src/renderers.ts#L64)

**Comment by master-q** [comment]
> do you mean it doesn't work in the classic notebook, or in JupyterLab?

classic notebook.

**Comment by fperez** [comment]
Just to clarify, without any ambiguity...

This code:

```python
import plotly.offline as py
py.init_notebook_mode(connected=False)

figure = {'data': [{'x': [0, 1], 'y': [0, 1], 'z': [0, 1], 'type': 'scatter3d'}],
          'layout': {'scene': {'xaxis': {'range': [0, 5], 'autorange': False},
                               'yaxis': {'range': [0, 5], 'autorange': False},
                               'zaxis': {'range': [0, 5], 'autorange': False}},
                     'updatemenus': [{'type': 'buttons',
                                      'buttons': [{'label': 'Play',
                                                   'method': 'animate',
                                                   'args': [None]}]}]
                    },
          'frames': [{'data': [{'x': [1, 2], 'y': [1, 2], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [1, 4], 'y': [1, 4], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [3, 4], 'y': [3, 4], 'z': [0, 3], 'type': 'scatter3d'}]}]
         }

py.iplot(figure)```

Is valid. When run in JupyterLab 0.31.12, it produces a static image with a non-functioning "Play" button:

![image](https://user-images.githubusercontent.com/57394/37577715-038f0b7a-2af1-11e8-800a-72bdab3fba7c.png)

and this error message in the console, just as @jasongrout suggested:

![image](https://user-images.githubusercontent.com/57394/37577719-134e7c6c-2af1-11e8-8153-044826aec688.png)

If the same notebook is opened in Classic (even within the same session), then the Play button animates as expected.

In summary, the animation issue seems indeed to be due to the JS output blocking. I don't know enough about how that JS policy works to understand why the animations are blocked but things like sliders (which I used in class on Thursday) do work fine.

**Comment by gnestor** [comment]
In JupyterLab, you would need to install the plotly-extension from [jupyter-renderers](https://github.com/jupyterlab/jupyter-renderers): `jupyter labextension install @jupyterlab/plotly-extension`

I have tried the below code with plotly-extension installed but the play button is still not animating...

```py
import plotly.offline as py

figure = {'data': [{'x': [0, 1], 'y': [0, 1], 'z': [0, 1], 'type': 'scatter3d'}],
          'layout': {'scene': {'xaxis': {'range': [0, 5], 'autorange': False},
                               'yaxis': {'range': [0, 5], 'autorange': False},
                               'zaxis': {'range': [0, 5], 'autorange': False}},
                     'updatemenus': [{'type': 'buttons',
                                      'buttons': [{'label': 'Play',
                                                   'method': 'animate',
                                                   'args': [None]}]}]
                    },
          'frames': [{'data': [{'x': [1, 2], 'y': [1, 2], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [1, 4], 'y': [1, 4], 'z': [0, 3], 'type': 'scatter3d'}]},
                     {'data': [{'x': [3, 4], 'y': [3, 4], 'z': [0, 3], 'type': 'scatter3d'}]}]
         }

py.iplot(figure)
```

Notice that `py.init_notebook_mode(connected=False)` is removed because JS output is disabled in lab.

**Comment by fperez** [comment]
Yup! @gnestor I should have clarified that I had the `plotly-extension` installed when I did the above...

**Comment by fperez** [comment]
Quick q - @gnestor @jasongrout, should we open a new issue to track the specific problem of animated output not working in JLab? I'm happy to do it and try to summarize the main points from the discussion above, but I'd like to ask in which repo we should do that. Main Lab one or jupyter-renderers?

**Comment by gnestor** [comment]
I'd say jupyter-renderers because it appears to be an extension-specific issue

**Comment by fperez** [comment]
[Done!](https://github.com/jupyterlab/jupyter-renderers/issues/118).  

I'm going to close this here as with current versions of plotly, all the animations work fine in Classic. In the issue above we can work out what's left for JLab via the renderers extension.

**Comment by harveyslash** [comment]
The 'animation' just looks like a series of 3 images. Are there any options to enable to get the points to move smoothly? 
in the 2d case, the points are smoothly transitioning by default
